Implement Higher-order expectation values in `GaussianState` using Wick's theorem

q-inho opened the pull request:
Resolve #437 
- I use Wick's theorem to implement `get_xp_string_expectation_value`

$$
\boxed{
\left\langle Y_{\alpha_1}Y_{\alpha_2}\cdots Y_{\alpha_n}\right\rangle
=\sum_{\substack{\text{all pairings}\\ \text{and singletons}}}
\Biggl(
    \prod_{\text{single }r}
      \mu_{\alpha_r}
\Biggr)
\Biggl(
    \prod_{(i,j)}
        \Sigma_{\alpha_i\alpha_j}
      +\frac{i\hbar}{2}
        \varepsilon_{\alpha_i\alpha_j}
\Biggr)
}
$$


Here $Y_{\alpha_i}$ are quadrature operators in xxpp ordering, $\mu$ is the mean vector, $\Sigma$ the covariance matrix, and $\epsilon_{\alpha_i \alpha_j} = \pm 1, 0$ encodes via $[x_i, p_i] = i\hbar$


- Two test codes are added `test_get_xp_string_expectation_value_against_fock_states` and `test_get_xp_string_expectation_value_random`.

- Edited: $+\mu_{\alpha_i} \mu_{\alpha_j}$ is removed from equation since I noticed I made a mistake. However, no change required from the code implementation side.

q-inho commented on piquasso/_simulators/gaussian/state.py:
```diff
@@ -960,6 +960,55 @@ def get_marginal_fock_probabilities(self, modes: Tuple[int, ...]) -> np.ndarray:
+        The indices of ``operators`` correspond to the xxpp ordering, i.e.
+        ``0`` denotes :math:`x_0`, ``d`` denotes :math:`p_0`, etc.
+
+        The implementation uses Wick's (Isserlis's) theorem.
```
Is Wikipedia is still good enough reference for this comment? Or should I add original paper here.

q-inho replied:
Added

q-inho commented on tests/_simulators/gaussian/test_state.py:
```diff
@@ -1021,3 +1021,85 @@ def test_get_purity_Thermal_Interferometer(generate_unitary_matrix):
+    return np.trace(rho @ operator)
+
+
+def test_get_xp_string_expectation_value_against_fock_states():
```
I have added single mode for analytic result. Howeve, am not confimdent with the analytic formulas for even 2 mode. If we need even 2 mod for testing with analyic, may I know which analytic formulas should I use?

Kolarovszki replied:
Here is an idea: a relative easy cross-check could be to consider pure Gaussian states $|\psi \rangle$ on $d$ modes, prepared from the vacuum by Gaussian gates:
```math
\langle \psi|  Y_1 Y_2 \dots Y_n |\psi \rangle = \langle 0| U^\dagger Y_1 Y_2 \dots Y_n U | 0 \rangle = \langle 0| U^\dagger Y_1 U U^\dagger Y_2 U \dots U_\dagger Y_n U | 0 \rangle.
```
We know that
```math
U^\dagger \mathbf{\chi}_i U =  \sum_{j=1}^{2d} S_{ij} \mathbf{\chi}_{j},
```
where $\mathbf{\chi} = [\hat{x}_1, \dots, \hat{x}_d, \hat{p}_1, \dots, \hat{p}_d]$ is the formal vector of the quadrature operators in the xxpp order, and $S \in \mathbb{C}^{2n \times 2n}$ is a symplectic matrix in the complex form. This is roughly described in Section 3.2.3 and 3.2.4 of [https://arxiv.org/pdf/1401.4679](https://arxiv.org/pdf/1401.4679). With these, one should be able to compute the moments for simple examples on paper I think.

q-inho commented:
@Kolarovszki Thank you for feedback. Actually I was quite unsure whether I should cite the [Isserlis's theorem](https://en.wikipedia.org/wiki/Isserlis%27s_theorem), since it did not exactly mentioned the derived equation. I tried to search whether other resources has exactly this equation, but I choose this while initially I thought it guide similar method like weyl order and operator contration.

So these are the simple derivations from [Isserlis's theorem](https://en.wikipedia.org/wiki/Isserlis%27s_theorem)
Equation from Isserlis's theorem
<img width="590" alt="image" src="https://github.com/user-attachments/assets/e13cf7c9-adf4-4693-9b04-fb40b61981ae" />

For classical centred theorem, Let $(X_1, \dots, X_n)$ be centred jointly Gaussian. For even $n$ and $P^2_n =$ {perfect pairing of { $1, \dots, n$ }}
<img width="264" alt="image" src="https://github.com/user-attachments/assets/addfb8e8-7e7b-4ac1-a66e-ffd3ebfb54b3" />

Let $X_k = \mu_k \Delta X_k$ with $\mathbb{E}[\Delta X_k] = 0$ and apply it to each block in the multinomial expansion of $\prod_k(\mu_k + \Delta X_k)$.
So for arbitrary $n$, it gives
<img width="385" alt="image" src="https://github.com/user-attachments/assets/ae8c499e-d362-4dc1-a9cd-9687d96715d8" />

Let $Y = (Y_1, \dots , Y_n)$ be Hermitian operators in a Gaussian state, The first and secomd moments are
<img width="625" alt="image" src="https://github.com/user-attachments/assets/c0120537-7c9e-4c1f-ade3-e5a5a79bf3e3" />


The Wey-ordered characteristic function is
<img width="309" alt="image" src="https://github.com/user-attachments/assets/797223f5-a2ce-4bc8-9ba0-33e8cff31b82" />

Define the unsymmetrized contraction
<img width="381" alt="image" src="https://github.com/user-attachments/assets/6bb2a74d-4944-4bf3-838d-73e185449eb8" />

Because $\log \chi_W$ is at most quadratic, all quantum cumulatns of order $> 2$ vanish. Conseqeuntly every multivariate derivative of $\chi_W$ obeys the same combinatorics as in the classical centred case, once the classical covariance $\Sigma_{ij}$ is replaced by the operator contraction $C_{ij}$. So the tracing it becomes
<img width="327" alt="image" src="https://github.com/user-attachments/assets/1177f0f2-113f-4de7-85ed-84c9b8e7ee33" />
that is valid for ordering of the plain product $Y_1 \cdots Y_n$.

Please let me know if this derivation leads to wrong methods.

Regarding the testing tolerance, I can push it to `tol=5e-3`. According to my impression I initially thought it was due to the finite cutoff in the Fock simulators that the xp operators are represented by truncated matrices whose commutator deviates from the canonical value. However, I am testing whether this statement is valid and cuase these high tolerance differences.

Kolarovszki commented:
@q-inho I also expected to find it written down somewhere, but I haven't managed either, so we have to write down the derivation ourselves instead. Sorry about this.

I think you're right, but your derivation skips many steps.

The step where you replace $\Sigma$ with $C$ is intuitive but nontrivial I think. This is how I see it. Suppose that the characteristic function giving the Weyl ordering of the quadrature operators is $`\chi_W(\mathbf{x}, \mathbf{p})`$. Then, writing the Weyl operator in terms of $\hat{x}_i$ and $`\hat{p}_i`$ and using BCH, we get that for normal ordering ($`\hat{x}_i`$ always in the front) we need to use $`\chi_W(\mathbf{x}, \mathbf{p}) e^{i \mathbf{x}\cdot\mathbf{p}}`$, which induces a shift in the covariance matrix $`\Sigma_{\text{normal}} := \Sigma + \frac{i}{2}X`$, where
```math
X = \begin{bmatrix}
    0 & I \\
    I & 0
\end{bmatrix}.
```
Of course this is known (https://journals.aps.org/pr/abstract/10.1103/PhysRev.177.1857). If you were to use $\Sigma_{\text{normal}}$ as the covariance matrix in Isserlis' formula, you would get the value for the normal ordered moment, but here terms of the form $`\langle \hat{p}_i \hat{x}_i \rangle`$ don't appear (which is also equal to $`\langle \hat{x}_i \hat{p}_i \rangle`$). Therefore, we can just insert $`\langle \hat{p}_i \hat{x}_i \rangle' = \langle \hat{x}_i \hat{p}_i \rangle - i \hbar`$. This yields a formula which works for normal ordered moments, but also incorporates the commutation relations, therefore it must work for any ordering. This is a neat trick.

Also, I agree with the characteristic function, but how do you differentiate to get the moments? What is $J$ here? According to Section 4.4 in Quantum Continuous Variables by Alessio Serafini, the covariance matrix in the characteristic function includes the symplectic form:
![image](https://github.com/user-attachments/assets/0192f127-c57a-4e5d-92df-55c2e8a500a0)
This flips $\hat{x}_i \mapsto - \hat{p}_i$ and $\hat{p}_i \mapsto \hat{x}_i$. This is because the one-mode Weyl operator has the form
```math
D(x, p) = \exp ( i \sqrt{2} ( x \hat{p} - p \hat{x}) ) =\exp ( i  \sqrt{2} x \hat{p}) \exp( - \sqrt{2} i p \hat{x}  )  \exp(i x p)
```
so differentiating by $x$ brings down $\hat{p}$ and by $-p$ it brings down $\hat{x}$. This is something one has to clarify during the derivation I think.

Regarding the tolerances: you're right, truncation of the observables can definitely lead to incorrect results, even if we increase the cutoff. At this moment, I don't know what to do about this -- I just mentioned that relying on these tests isn't a safe bet, as you could easily get a similar result within that tolerance by reordering the operators.

q-inho commented:
@Kolarovszki Thanks. First of all, I changed the reference for the algorithm to the Wikipedia (for now) while I am not sure whether I should provide full derivation on the comment of the function. 

For the comparing test with Fock state, I leave it with tol 1e-2, but left the comment that that test file is not a safe bet.

Regarding the analytic tests, I added tests that validate `get_xp_string_expectation_value` against an analytic calculation. The helper `analytic_xp_string_expectation_value` computes moments directly from the symplectic matrix $S$ and displacement $d$ using the relation $U^\dagger \chi_i U = \sum_j S_{ij} \chi_j$. The tests cover both displaced and nondisplaced single‑mode cases, as well as a nondisplaced two‑mode state, ensuring higher‑order moments match the analytic predictions. Please let me know if you’d like additional examples or modifications.

Kolarovszki commented:
> @Kolarovszki Thanks. First of all, I changed the reference for the algorithm to the Wikipedia (for now) while I am not sure whether I should provide full derivation on the comment of the function.
> 
> For the comparing test with Fock state, I leave it with tol 1e-2, but left the comment that that test file is not a safe bet.
> 
> Regarding the analytic tests, I added tests that validate `get_xp_string_expectation_value` against an analytic calculation. The helper `analytic_xp_string_expectation_value` computes moments directly from the symplectic matrix S and displacement d using the relation U † χ i U = ∑ j S i j χ j . The tests cover both displaced and nondisplaced single‑mode cases, as well as a nondisplaced two‑mode state, ensuring higher‑order moments match the analytic predictions. Please let me know if you’d like additional examples or modifications.

Thank you! I have minor suggestions, but I don't want to burden you with this as you have already worked so much on this, is it good for you if I add commits to your PR?

Regarding the source of this formula, I think it would be good to write down the argument in the docstring, but I can also do it.

q-inho commented:
@Kolarovszki Thanks a lot! Yes, of course you can add commits to my PR. I really learn a lot through this PR work.

Kolarovszki reviewed (approved):
Thank you for your contribution!
